src/Processors/Transforms/WindowTransform.cpp

```diff
@@ -2154,6 +2156,101 @@ struct WindowFunctionNthValue final : public WindowFunction
 };
 
 
+// nonNegativeDerivative(metric_column, timestamp_column[, INTERVAL 1 SECOND])
+struct WindowFunctionNonNegativeDerivative final : public RecurrentWindowFunction<0>
+{
+    static constexpr size_t ARGUMENT_METRIC = 0;
+    static constexpr size_t ARGUMENT_TIMESTAMP = 1;
+    static constexpr size_t ARGUMENT_INTERVAL = 2;
+
+    WindowFunctionNonNegativeDerivative(const std::string & name_,
+                                            const DataTypes & argument_types_, const Array & parameters_)
+        : RecurrentWindowFunction(name_, argument_types_, parameters_)
+    {
+        if (!parameters.empty())
+        {
+            throw Exception(ErrorCodes::BAD_ARGUMENTS,
+                            "Function {} cannot be parameterized", name_);
+        }
+
+        if (argument_types.size() != 2 && argument_types.size() != 3)
+        {
+            throw Exception(ErrorCodes::BAD_ARGUMENTS,
+                            "Function {} takes 2 or 3 arguments", name_);
+        }
+
+        if (!isNumber(argument_types[ARGUMENT_METRIC]))
+        {
+            throw Exception(ErrorCodes::BAD_ARGUMENTS,
+                            "Argument {} must be a number, '{}' given",
+                            ARGUMENT_METRIC,
+                            argument_types[ARGUMENT_METRIC]->getName());
+        }
+
+        if (!isDateTime(argument_types[ARGUMENT_TIMESTAMP]) && !isDateTime64(argument_types[ARGUMENT_TIMESTAMP]))
+        {
+            throw Exception(ErrorCodes::BAD_ARGUMENTS,
+                            "Argument {} must be DateTime or DateTime64, '{}' given",
+                            ARGUMENT_TIMESTAMP,
+                            argument_types[ARGUMENT_TIMESTAMP]->getName());
+        }
+
+        if (argument_types.size() == 3)
+        {
+            interval_specified = true;
+            const DataTypeInterval * interval_datatype = checkAndGetDataType<DataTypeInterval>(argument_types[ARGUMENT_INTERVAL].get());
+            if (!interval_datatype)
+            {
+                throw Exception(
+                    ErrorCodes::BAD_ARGUMENTS,
+                    "Argument {} must be an INTERVAL, '{}' given",
+                    ARGUMENT_INTERVAL,
+                    argument_types[ARGUMENT_INTERVAL]->getName());
+            }
+            if (interval_datatype->getKind() == IntervalKind::Month || interval_datatype->getKind() == IntervalKind::Quarter
```

Comment:
Perhaps wrap this condition in a method IntervalKind::hasNonConstSeconds()?

Reply:
Good idea. Willdo